simulation/modules/macrophage.py

```diff
@@ -157,3 +221,75 @@ def valid_point(point: Point, grid: RectangularGrid):
         cells.update_voxel_index([index])
 
     return cells
+
+
+def hillProbability(substract, km=10):
+    return substract * substract / (substract * substract + km * km)
+
+
+def logistic(x, l, b):
+    return 1 - b * math.exp(-((x / l) ** 2))
+
+
+def chemotaxis(
+    molecule, P, drift_lambda, drift_bias, cells: MacrophageCellList, tissue, grid: RectangularGrid
+):
+    # 'molecule' = state.'molecule'.concentration
+    # P = some probability to move
+
+    # 1. Get cells that are alive
+    for index in cells.alive():
+
+        # 2. Get voxel for each cell to get molecule in that voxel
+        cell = cells[index]
+        vox = grid.get_voxel(cell['point'])
+
+        # 3. Set prob for neighboring voxels
+        p = []
+        vox_list = []
+        p_tot = 0
+        i = -1
+
+        # calculate individual probability
+        for x in [0, 1, -1]:
+            for y in [0, 1, -1]:
+                for z in [0, 1, -1]:
+                    p.append(0.0)
+                    vox_list.append([x, y, z])
+                    i += 1
+                    if (
+                        (grid.xv[0] <= x)
+                        & (x <= grid.xv[-1])
+                        & (grid.yv[0] <= y)
+                        & (y <= grid.yv[-1])
+                        & (grid.zv[0] <= z)
+                        & (z <= grid.zv[-1])
+                    ):
+                        voxel = Voxel(z=(vox.z + z), y=(vox.y + y), x=(vox.x + x))
+                        if tissue[voxel.z, voxel.y, voxel.x] == (
+                            TissueTypes.SURFACTANT.value
+                            or TissueTypes.BLOOD.value
+                            or TissueTypes.EPITHELIUM.value
+                            or TissueTypes.PORE.value
+                        ):
+                            p[i] = logistic(
+                                molecule[voxel.z, voxel.y, voxel.x], drift_lambda, drift_bias
+                            )
+                            p_tot += p[i]

```

Comment:
A couple of things about this loop:
* Take a look at the grid documentation I wrote: https://nutritionallungimmunity.org/simulation-framework/grid.html  I'm hoping it will help clear up the confusion.  To help see what you are doing wrong here, try renaming the `x`, `y`, and `z` variables to `i`, `j`, and `k`.  This is how they are represented in the documentation.
* I think we will want to add a function to the voxel api to return an iterator over all of the neighbors.  That will simplify this code and minimize breakage if we want to migrate to an unstructured grid.

Reply:
To be honest I am still a little confused exactly what I am doing wrong. I am hoping to access the concentration of the molecule in a voxel neighborhood of a voxel that contains a cell's point location. I thought that by first using 'get_voxel'  for a point that I am getting the 'center' for the grid cell containing that point and then by adding one to that center I am getting the center of the neighboring grid cell. I'm sorry, I'm not sure why this has been such a difficult concept for me to grasp.

I agree, an iterator of a voxel's neighbors will be very useful and commonly used.

Reply:
I think you are confused by at least two different things.  At the very least, you need to keep track of the difference between an index (an integer) and a coordinate/position (a float).

> I thought that by first using 'get_voxel' for a point that I am getting the 'center' for the grid cell containing that point

This is partially my fault in conflating the `Point` and `Voxel` classes.  Clearly the docstrings need to be more clear here.  For a voxel, `vox.x` is the **index** of the the voxel not the coordinate of the center.  So in the code when you do essentially `grid.xv[0] <= vox.x + 1`, you are comparing and index with a coordinate.

> by adding one to that center I am getting the center of the neighboring grid cell. 

If the attributes of the voxel class did represent the coordinates of the cell center, this would still be wrong because the distance between **coordinates** of adjacent cells is not generally 1.

The major logical error in this loop is the comparison `grid.xv[0] <= xi <= grid.xv[-1]`.  I think you want to ensure the indices are not out of bounds, so the correct expression is `0 <= xi <= grid.shape[2] - 1` because `xi` is an index value, not a coordinate.